File: ecos-records/src/main/java/ru/citeck/ecos/records2/graphql/meta/value/factory/bean/BeanValueFactory.java
package ru.citeck.ecos.records2.graphql.meta.value.factory.bean;

import lombok.extern.slf4j.Slf4j;
import org.apache.commons.beanutils.PropertyUtils;
import org.jetbrains.annotations.NotNull;
import ru.citeck.ecos.commons.data.DataValue;
import ru.citeck.ecos.commons.json.Json;
import ru.citeck.ecos.commons.utils.ExceptionUtils;
import ru.citeck.ecos.records2.RecordRef;
import ru.citeck.ecos.records2.graphql.meta.value.MetaEdge;
import ru.citeck.ecos.records2.graphql.meta.value.MetaField;
import ru.citeck.ecos.records2.graphql.meta.value.MetaValue;
import ru.citeck.ecos.records2.graphql.meta.value.SimpleMetaEdge;
import ru.citeck.ecos.records2.graphql.meta.value.factory.MetaValueFactory;
import ru.citeck.ecos.records2.graphql.meta.value.field.EmptyMetaField;

import java.beans.PropertyDescriptor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;

@Slf4j
public class BeanValueFactory implements MetaValueFactory<Object> {

    @Override
    public MetaValue getValue(Object value) {
        return new Value(value);
    }

    @Override
    public List<Class<?>> getValueTypes() {
        return Collections.singletonList(Object.class);
    }

    static class Value implements MetaValue {

        private final Object bean;
        private final BeanTypeContext typeCtx;

        Value(Object bean) {
            this.bean = bean;
            typeCtx = BeanTypeUtils.getTypeContext(bean.getClass());
        }

        @Override
        public String getId() {
            if (typeCtx.hasProperty(".id")) {
                return getAttWithType(".id", String.class);
            }
            try {
                Object id = PropertyUtils.getProperty(bean, "id");
                return id != null ? id.toString() : null;
            } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
                return null;
            }
        }

        @Override
        public Double getDouble() {
            if (typeCtx.hasProperty(".num")) {
                return getAttWithType(".num", Double.class);
            }
            String str = getString();
            return str != null ? Double.parseDouble(str) : null;
        }

        @Override
        public Boolean getBool() {
            if (typeCtx.hasProperty(".bool")) {
                return getAttWithType(".bool", Boolean.class);
            }
            return Boolean.parseBoolean(getString());
        }

        @Override
        public RecordRef getRecordType() {
            return getAttWithType(".type", RecordRef.class);
        }

        @Override
        public String getDisplayName() {
            if (typeCtx.hasProperty(".disp")) {
                return getAttWithType(".disp", String.class);
            }
            return getString();
        }

        @Override
        public String getString() {
            if (typeCtx.hasProperty(".str")) {
                return getAttWithType(".str", String.class);
            }
            return bean.toString();
        }

        @Override
        public Object getAttribute(@NotNull String name, @NotNull MetaField field) throws Exception {
            if (bean instanceof Map) {
                @SuppressWarnings("unchecked")
                Map<String, Object> beanAsMap = (Map<String, Object>) bean;
                return beanAsMap.get(name);
            }
            return typeCtx.getProperty(bean, name);
        }

        @Override
        public boolean has(@NotNull String name) {
            return typeCtx.hasProperty(name);
        }

        @Override
        public Object getJson() {
            if (typeCtx.hasProperty(".json")) {
                return getAttWithType(".json", DataValue.class);
            }
            return Json.getMapper().toJson(bean);
        }

        Object getBean() {
            return bean;
        }

        @Override
        public MetaEdge getEdge(@NotNull String name, @NotNull MetaField field) {
            return new BeanEdge(name, this);
        }

        private <T> T getAttWithType(String name, Class<T> type) {
            try {
                return Json.getMapper().convert(getAttribute(name, EmptyMetaField.INSTANCE), type);
            } catch (Exception e) {
                return null;
            }
        }
    }

    static class BeanEdge extends SimpleMetaEdge {

        private final Object bean;
        private PropertyDescriptor descriptor;

        BeanEdge(String name, Value scope) {
            super(name, scope);
            bean = scope.getBean();
        }

        @Override
        public boolean isMultiple() {
            return Collection.class.isAssignableFrom(getDescriptor().getPropertyType());
        }

        @Override
        public Class<?> getJavaClass() {

            PropertyDescriptor descriptor = getDescriptor();

            if (descriptor == null) {
                return null;
            }

            Class<?> type = descriptor.getPropertyType();

            if (Collection.class.isAssignableFrom(type)) {

                Type returnType = descriptor.getReadMethod().getGenericReturnType();

                ParameterizedType parameterType = (ParameterizedType) returnType;
                return (Class<?>) parameterType.getActualTypeArguments()[0];
            }

            return type;
        }

        private PropertyDescriptor getDescriptor() {
            if (descriptor == null) {
                try {
                    descriptor = PropertyUtils.getPropertyDescriptor(bean, getName());
                } catch (NoSuchMethodException e) {
                    log.debug("Descriptor not found", e);
                } catch (IllegalAccessException | InvocationTargetException e) {
                    ExceptionUtils.throwException(e);
                    throw new RuntimeException(e); // never executed
                }
            }
            return descriptor;
        }
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/atts/value/factory/bean/BeanValueFactory.kt
package ru.citeck.ecos.records3.record.atts.value.factory.bean

import org.apache.commons.beanutils.PropertyUtils
import ru.citeck.ecos.commons.data.DataValue
import ru.citeck.ecos.commons.json.Json
import ru.citeck.ecos.records2.RecordConstants
import ru.citeck.ecos.records3.record.atts.schema.ScalarType
import ru.citeck.ecos.records3.record.atts.value.AttEdge
import ru.citeck.ecos.records3.record.atts.value.AttValue
import ru.citeck.ecos.records3.record.atts.value.factory.AttValueFactory
import ru.citeck.ecos.records3.record.atts.value.impl.SimpleAttEdge
import java.beans.PropertyDescriptor
import java.lang.reflect.ParameterizedType

class BeanValueFactory : AttValueFactory<Any> {

    override fun getValue(value: Any): AttValue {
        return Value(value)
    }

    override fun getValueTypes() = listOf(Any::class.java)

    internal class Value(val bean: Any) : AttValue {

        private val typeCtx: BeanTypeContext = BeanTypeUtils.getTypeContext(bean.javaClass)

        override fun getId(): Any? {
            if (typeCtx.hasProperty(ScalarType.ID.schema)) {
                return getAttWithType(ScalarType.ID.schema, Any::class.java)
            }
            return null
        }

        override fun asDouble(): Double? {
            if (typeCtx.hasProperty(ScalarType.NUM.schema)) {
                return getAttWithType(ScalarType.NUM.schema, Double::class.java)
            }
            return asText()?.toDouble()
        }

        override fun asBoolean(): Boolean? {
            if (typeCtx.hasProperty(ScalarType.BOOL.schema)) {
                return getAttWithType(ScalarType.BOOL.schema, Boolean::class.java)
            }
            return asText()?.toBoolean()
        }

        override fun getType(): Any? {
            if (typeCtx.hasProperty(RecordConstants.ATT_TYPE)) {
                return getAttWithType(RecordConstants.ATT_TYPE, Any::class.java)
            }
            return null
        }

        override fun getDisplayName(): Any? {
            if (typeCtx.hasProperty(ScalarType.DISP.schema)) {
                return getAttWithType(ScalarType.DISP.schema, Any::class.java)
            }
            return asText()
        }

        override fun asText(): String? {
            if (typeCtx.hasProperty(ScalarType.STR.schema)) {
                return getAttWithType(ScalarType.STR.schema, String::class.java)
            }
            return bean.toString()
        }

        @Throws(Exception::class)
        override fun getAtt(name: String): Any? {
            if (bean is Map<*, *>) {
                return bean[name]
            }
            return typeCtx.getProperty(bean, name)
        }

        override fun has(name: String): Boolean {
            if (bean is Map<*, *>) {
                return bean.containsKey(name)
            }
            return typeCtx.beanHas(bean, name)
        }

        override fun getAs(type: String): Any? {
            return typeCtx.beanGetAs(bean, type)
        }

        override fun asJson(): Any? {
            if (typeCtx.hasProperty(ScalarType.JSON.schema)) {
                return getAttWithType(ScalarType.JSON.schema, DataValue::class.java)
            }
            return Json.mapper.toJson(bean)
        }

        override fun asRaw(): Any? {
            if (typeCtx.hasProperty(ScalarType.RAW.schema)) {
                return getAttWithType(ScalarType.RAW.schema, Any::class.java)
            }
            return Json.mapper.toJson(bean)
        }

        override fun getEdge(name: String): AttEdge {
            return typeCtx.beanGetEdge(bean, name) ?: BeanEdge(name, this)
        }

        private fun <T : Any> getAttWithType(name: String, type: Class<T>): T? {
            val value = getAtt(name)
            return if (type != Any::class.java) {
                Json.mapper.convert(getAtt(name), type)
            } else {
                @Suppress("UNCHECKED_CAST")
                value as? T
            }
        }
    }

    internal class BeanEdge(name: String, scope: Value) : SimpleAttEdge(name, scope) {

        // never executed
        private val descriptor: PropertyDescriptor? by lazy {
            try {
                PropertyUtils.getPropertyDescriptor(scope.bean, getName())
            } catch (e: NoSuchMethodException) {
                log.debug("Descriptor not found", e)
                null
            }
        }

        override fun isMultiple(): Boolean {
            val desc = descriptor ?: return false
            return Collection::class.java.isAssignableFrom(desc.propertyType)
        }

        override fun getJavaClass(): Class<*>? {
            val descriptor = descriptor ?: return null
            val type = descriptor.propertyType
            if (MutableCollection::class.java.isAssignableFrom(type)) {
                val returnType = descriptor.readMethod.genericReturnType
                val parameterType = returnType as ParameterizedType
                return parameterType.actualTypeArguments[0] as? Class<*>
            }
            return type
        }
    }
}
